src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinderProvider.cs

```diff
@@ -21,26 +20,36 @@ public IModelBinder GetBinder(ModelBinderProviderContext context)
                 throw new ArgumentNullException(nameof(context));
             }
 
-            if (context.BindingInfo.BindingSource != null &&
-                context.BindingInfo.BindingSource.CanAcceptDataFrom(BindingSource.Header))
+            var modelMetadata = context.Metadata;
+
+            if (context.BindingInfo.BindingSource != null
+                && context.BindingInfo.BindingSource.CanAcceptDataFrom(BindingSource.Header)
+                && IsSimpleType(modelMetadata))
             {
-                var loggerFactory = context.Services.GetRequiredService<ILoggerFactory>();
-                var logger = loggerFactory.CreateLogger<HeaderModelBinderProvider>();
+                var metadata = modelMetadata.GetMetadataForType(modelMetadata.ModelType);
 
-                // We only support strings and collections of strings. Some cases can fail
-                // at runtime due to collections we can't modify.
-                if (context.Metadata.ModelType == typeof(string) ||
-                    context.Metadata.ElementType == typeof(string))
-                {
-                    return new HeaderModelBinder(loggerFactory);
-                }
-                else
+                // Change the binding info to prevent recursion
```

Comment:
This is probably worth a bit more explanation given how long we discussed it today :)